kernel/pci/src/lib.rs

```diff
@@ -211,84 +249,133 @@ pub struct PciLocation {
+    byte_shift: u32,
+}
+
 impl PciLocation {
     pub fn bus(&self) -> u8 { self.bus }
     pub fn slot(&self) -> u8 { self.slot }
     pub fn function(&self) -> u8 { self.func }
 
 
-    /// Computes a PCI address from bus, slot, func, and offset. 
-    /// The least two significant bits of offset are masked, so it's 4-byte aligned addressing,
-    /// which makes sense since we read PCI data (from the configuration space) in 32-bit chunks.
-    fn pci_address(self, offset: u8) -> u32 {
-        ((self.bus  as u32) << 16) | 
-        ((self.slot as u32) << 11) | 
-        ((self.func as u32) <<  8) | 
-        ((offset as u32) & (PCI_CONFIG_ADDRESS_OFFSET_MASK as u32)) | 
-        0x8000_0000
+    /// Computes a [`PciAddress`] from bus, slot, func, and offset. 
+    ///
+    /// PCI configuration space addresses must be aligned to 4 bytes (double word)
+    /// for reads and writes to succeed. The `dword_address` field of the returned
+    /// structure will be aligned to 4 bytes accordingly.
+    ///
+    /// Offsets with the two least significant bits set are still valid (they point
+    /// to words and bytes). This function computes a `byte_shift` value which should
+    /// be used to shift the double word pointed to by `dword_address`.
+    fn pci_address_and_shift(self, offset: u8) -> PciAddress {
+        let dword_offset = (offset & PCI_CONFIG_ADDRESS_OFFSET_MASK) as u32;
+        let byte_shift = ((offset & (!PCI_CONFIG_ADDRESS_OFFSET_MASK)) * 8) as u32;
+
+        let dword_address = BASE_OFFSET
+            | ((self.bus  as u32) << 16)
+            | ((self.slot as u32) << 11)
+            | ((self.func as u32) <<  8)
+            | dword_offset;
+
+        PciAddress {
+            dword_address,
+            byte_shift,
+        }
     }
 
-    /// read 32-bit data at the specified `offset` from the PCI device specified by the given `bus`, `slot`, `func` set.
+    /// Read 32-bit data at the specified `offset` from this PCI device.
     fn pci_read_32(&self, offset: u8) -> u32 {
-        unsafe { 
-            PCI_CONFIG_ADDRESS_PORT.lock().write(self.pci_address(offset)); 
+        let address = self.pci_address_and_shift(offset);
+        let value;
+
+        #[cfg(target_arch = "x86_64")] {
+            unsafe { 
+                PCI_CONFIG_ADDRESS_PORT.lock().write(address.dword_address);
+            }
+            value = PCI_CONFIG_DATA_PORT.lock().read();
         }
-        Self::read_data_port() >> ((offset & (!PCI_CONFIG_ADDRESS_OFFSET_MASK)) * 8)
+
+        #[cfg(target_arch = "aarch64")] {
+            let config_space = PCI_CONFIG_SPACE.lock();
+            let config_space = config_space.get()
+                .expect("PCI Config Space wasn't mapped yet");
+            value = unsafe {
+                config_space
+                    .as_ptr()
+                    .add(address.dword_address as usize)
+                    .cast::<u32>()
+                    .read_volatile()
```

Comment:
hmm, so if we map the config space as an array (i.e., `BorrowedSliceMappedPages`), why do we need to use unsafe pointer arithmetic here instead of just indexing into the array?

Seems like you ought to be able to do something like map the config space as a slice of `&mut [Volatile<u32>]` and then access it by indexing into it with the index being `address.dword_address / 4`.

As a bonus, this would also avoid unaligned reads/writes.

Reply:
Indeed; will do